lightning/src/offers/invoice.rs

```diff
@@ -0,0 +1,1543 @@
+enum InvoiceContents {
+	/// Contents for an [`Invoice`] corresponding to an [`Offer`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	ForOffer {
+		invoice_request: InvoiceRequestContents,
+		fields: InvoiceFields,
+	},
+	/// Contents for an [`Invoice`] corresponding to a [`Refund`].
+	///
+	/// [`Refund`]: crate::offers::refund::Refund
+	ForRefund {
+		refund: RefundContents,
+		fields: InvoiceFields,
+	},
+}
+
+/// Invoice-specific fields for an `invoice` message.
+struct InvoiceFields {
+	payment_paths: Vec<(BlindedPath, BlindedPayInfo)>,
+	created_at: Duration,
+	relative_expiry: Option<Duration>,
+	payment_hash: PaymentHash,
+	amount_msats: u64,
+	fallbacks: Option<Vec<FallbackAddress>>,
+	features: Bolt12InvoiceFeatures,
+	signing_pubkey: PublicKey,
+}
+
+impl Invoice {
+	/// Paths to the recipient originating from publicly reachable nodes, including information
+	/// needed for routing payments across them. Blinded paths provide recipient privacy by
+	/// obfuscating its node id.
+	pub fn payment_paths(&self) -> &[(BlindedPath, BlindedPayInfo)] {
+		&self.contents.fields().payment_paths[..]
+	}
+
+	/// Duration since the Unix epoch when the invoice was created.
+	pub fn created_at(&self) -> Duration {
+		self.contents.fields().created_at
+	}
+
+	/// Duration since [`Invoice::created_at`] when the invoice has expired and therefore should no
+	/// longer be paid.
+	pub fn relative_expiry(&self) -> Duration {
+		self.contents.fields().relative_expiry.unwrap_or(DEFAULT_RELATIVE_EXPIRY)
+	}
+
+	/// Whether the invoice has expired.
+	#[cfg(feature = "std")]
+	pub fn is_expired(&self) -> bool {
+		let absolute_expiry = self.created_at().checked_add(self.relative_expiry());
+		match absolute_expiry {
+			Some(seconds_from_epoch) => match SystemTime::UNIX_EPOCH.elapsed() {
+				Ok(elapsed) => elapsed > seconds_from_epoch,
+				Err(_) => false,
+			},
+			None => false,
+		}
+	}
+
+	/// SHA256 hash of the payment preimage that will be given in return for paying the invoice.
+	pub fn payment_hash(&self) -> PaymentHash {
+		self.contents.fields().payment_hash
+	}
+
+	/// The minimum amount required for a successful payment of the invoice.
+	pub fn amount_msats(&self) -> u64 {
+		self.contents.fields().amount_msats
+	}
+
+	/// Fallback addresses for paying the invoice on-chain, in order of most-preferred to
+	/// least-preferred.
+	pub fn fallbacks(&self) -> Vec<Address> {
+		let network = match self.network() {
+			None => return Vec::new(),
+			Some(network) => network,
+		};
+
+		let to_valid_address = |address: &FallbackAddress| {
+			// This is redundant with Address::is_standard for V0 and V1 but is a spec requirement.
+			let program = &address.program;
+			if program.len() < 2 || program.len() > 40 {
+				return None;
+			}
+
+			let address = Address {
+				payload: Payload::WitnessProgram {
+					version: address.version,
+					program: address.program.clone(),
+				},
+				network,
+			};
+
+			if !address.is_standard() {
```

Comment:
Yeah, that does make sense. I may have been reading too much into this part of the spec:

> - MUST ignore any `fallback_address` for which `address` does not meet known requirements for the given `version`

Since there wouldn't be any known requirements for, say, version 3, we shouldn't ignore them, I guess?

Reply:
Yeah, length of V0 and V1 scripts are covered by `is_standard` but other versions are considered non-standard. Best to have it fixed upstream, I agree, but we'll have to explicitly allow for other versions now, it seems?

Reply:
Looks like V1 must be `WITNESS_V1_TAPROOT_SIZE` (32), which is what `Address::is_standard` checks.

Reply:
Oh, nevermind. I see the fallt-hrough case is fine with it.

Reply:
Ok, now we only skip V0 addresses that aren't the right size.